core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +120,146 @@ where
             hasher,
             hashes,
             binary_tree_size,
+            start_index: 0,
+            cache: vec![None; depth],
         }
     }
 
+    /// Returns `true` if the tree is empty.
+    pub fn is_empty(&self) -> bool {
+        self.start_index == 0 && self.hashes.is_empty()
+    }
+
     /// Returns the root hash of this tree.
     /// # Panics
-    /// Will panic if the constant below is invalid.
-    pub fn merkle_root(self) -> H256 {
+    /// Should never panic, unless there is a bug.
+    pub fn merkle_root(&self) -> H256 {
         if self.hashes.is_empty() {
             let depth = tree_depth_by_size(self.binary_tree_size);
-            self.hasher.empty_subtree_hash(depth)
-        } else {
-            self.compute_merkle_root_and_path(0, None)
+            if self.start_index == 0 {
+                return self.hasher.empty_subtree_hash(depth);
+            } else if self.start_index == self.binary_tree_size {
+                return self.cache[depth].expect("cache is invalid");
+            }
         }
+        self.compute_merkle_root_and_path(0, None, None)
     }
 
     /// Returns the root hash and the Merkle proof for a leaf with the specified 0-based `index`.
-    pub fn merkle_root_and_path(self, index: usize) -> (H256, Vec<H256>) {
-        let mut merkle_path = vec![];
-        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut merkle_path));
-        (root_hash, merkle_path)
+    /// `index` is relative to the leftmost uncached leaf.
+    /// # Panics
+    /// Panics if `index` is >= than the number of leaves in the tree.
+    pub fn merkle_root_and_path(&self, index: usize) -> (H256, Vec<H256>) {
+        assert!(index < self.hashes.len(), "leaf index out of bounds");
+        let mut end_path = vec![];
+        let root_hash = self.compute_merkle_root_and_path(index, Some(&mut end_path), None);
+        (
+            root_hash,
+            end_path.into_iter().map(Option::unwrap).collect(),
+        )
+    }
+
+    /// Returns the root hash and the Merkle proofs for a range of leafs.
+    /// The range is 0..length, where `0` is the leftmost untrimmed leaf.
+    /// # Panics
+    /// Panics if `length` is 0 or greater than the number of leaves in the tree.
+    pub fn merkle_root_and_paths_for_range(
+        &self,
+        length: usize,
+    ) -> (H256, Vec<Option<H256>>, Vec<Option<H256>>) {
```

Comment:
As a further suggestion to the previous comment: maybe, it makes sense to create a newtype for Merkle paths returned here?

```rust
pub struct MerklePath(Vec<Option<H256>>);

impl MerklePath {
    fn new(full_path: Vec<H256>, side: Side) -> Self {
        // ...
    }

    // needs some sort of a getter as well; not sure how exactly it should look
}
```